packages/core/database/lib/lifecycles/index.js

```diff
@@ -30,31 +30,39 @@ const createLifecyclesProvider = db => {
       subscribers = [];
     },
 
-    createEvent(action, uid, properties) {
+    createEvent(action, uid, properties, state) {
       const model = db.metadata.get(uid);
 
       return {
         action,
         model,
+        state,
         ...properties,
       };
     },
 
-    async run(action, uid, properties) {
-      for (const subscriber of subscribers) {
+    async run(action, uid, properties, states) {
```

Comment:
I don't understand the logic you use with states being considered as an array here? 
We want to pass the same state to all subscribers just passing a simple state object around would be enough I think

Reply:
Sorry to intrude, but I have a question.
Is the `state` object shared across all lifecycles or is it model-specific?

Reply:
From my understanding, using only one state object create the risk of a subscriber replacing the values put in the state by a previous subscriber, potentially causing issues for future events down the line.

Say we plug in two external libraries along with a couple of custom subscribers, I now have their states mixed with mine and I am at risk of replacing critical values by mistake. Of course this will most likely break the code of the events registered by the external libraries as the state doen't match their expectation anymore.

Using a different state for each subscriber identity prevent this issue and all events receive exactly the state they defined, no matter how many subscribers are registered.

I'll be happy to be proven wrong and to use a simple object rather than an array of objects, please let me know!

Reply:
Hi @rgoupil thank you for elaborating on the idea you had. It does make a lot of sense to scope the states per subscriber indeed 👍 I am still thinking a global state could be very useful but more for a middleware-like approach though. we could introduce that in the future.

That being said, I would still pass around a single "state" even if it contains some logic to distribute the right context to each subscriber. I think the lifecycle should be more "dumb" in that regard. Here is what I have in mind

```
const { state } = await db.lifecycles.run('beforeFindOne');
//...
 await db.lifecycles.run('afterFindOne', ..., state);
```

Where state would be a Map from subscriber to state. This would avoid race conditions that the array index approach can have.

Each lifecycle could return its state and initialize one if none was provided. Pretty close to your approach but more abstracted away from the entity manager.

Hope this makes sense :)

Reply:
It makes total sense, I'll do just that :)
I disliked the `Array` approach however I failed to remember that a `Map` can handle objects and functions as keys (as opposed to an `Object`), thanks for the reminder!

Reply:
I updated the implementation to reflect your changes.
Thanks for the review, I'll mark this conversation as resolved!